internal/controller/helpers.go

```diff
@@ -54,146 +53,75 @@ func fetchResource[T WSAResource](ctx context.Context, c client.Client, req ctrl
 	return resource, nil
 }
 
-func ensureFinalizer[T client.Object](ctx context.Context, c client.Client, resource T) bool {
-	log := logf.FromContext(ctx)
-
-	if addFinalizer(resource) {
-		if err := c.Update(ctx, resource); err != nil {
-			log.Error(err, "failed to add finalizer")
-			return false
-		}
-		log.Info("Added finalizer to resource")
-		return true
-	}
-	return false
-}
-
+// handleDeletion manages the deletion flow for WSA/cWSA resources.
+// It waits for staging to process the deletion before removing the finalizer.
+// Returns (requeue, error) where requeue indicates if the reconcile should be retried.
 func handleDeletion[T WSAResource](
 	ctx context.Context,
 	c client.Client,
-	req ctrl.Request,
+	engine *rules.InMemoryEngine,
 	resource T,
-	engine rules.Engine,
-	wsaResource rules.WSAResource,
-) (ctrl.Result, error) {
+) (bool, error) {
 	log := logf.FromContext(ctx)
 
-	log.Info("Entering deletion handler",
-		"name", resource.GetName(),
-		"namespace", resource.GetNamespace(),
-		"hasFinalizer", hasFinalizer(resource.GetFinalizers(), ServiceAccountCleanupFinalizer))
-
-	if !hasFinalizer(resource.GetFinalizers(), ServiceAccountCleanupFinalizer) {
-		log.V(1).Info("Resource being deleted but finalizer already removed")
-		return ctrl.Result{}, nil
+	if !containsCleanupFinalizer(resource) {
+		log.V(1).Info("Resource has no cleanup finalizer, allowing deletion",
+			"name", resource.GetName(),
+			"namespace", resource.GetNamespace())
+		return false, nil
 	}
 
-	log.Info("Cleaning up ServiceAccounts for deleted resource")
-
-	result, err := engine.CleanupServiceAccounts(ctx, wsaResource)
-	if err != nil {
-		log.Error(err, "failed to cleanup ServiceAccounts")
-		return result, err
+	// Use full namespace/name key to correctly identify the resource.
+	// For cluster-scoped resources (namespace is empty), this will just be the name.
+	wsaKey := resource.GetName()
+	if ns := resource.GetNamespace(); ns != "" {
+		wsaKey = ns + "/" + resource.GetName()
```

Comment:
This may not be helpful information, but this is functionality built into the `NamespacedNamed` type. Might be worth leveraging that to consolidate.